### ISSUE TITLE
Bug: Console logs stop updating when non-ascii characters are encountered

### ISSUE THREAD

**Issue by BrewRunner** [gocd/gocd#3926]
**Bug Report**

**Summary**

Console logs containing non-ASCII characters are truncated at the point that the non-ASCII characters occur. Only the portion of the console log that is generated prior to the occurrence of the non-ASCII characters is displayed. The log is still generated and view-able once the stage completes via the "raw output" link, however the non-ASCII characters are replaced with this symbol "�". 

We discovered this issue when upgrading from version 17.2 to version 17.10 and were forced to back out of the upgrade as a result of the inability to follow the progress of any stages that display non-ASCII characters in the console log.  Many of these stages are lengthy, causing the stage to appear stalled since the log is no longer updating and the "raw output" is not available until the stage completes.

The raw output in 17.2 also replaces non-ASCII characters with "�", but does process and display them correctly in the scrolling log.

The specific characters in our logs that may be contributing to the problem are: æ, ä, ö, ê, à, ü, ö, ß, ó, í, ç, á, 

**Environment**

GoCD 17.10
- non-ASCII characters cause console log to stop rendering "real time"
- non-ASCII characters are replaced by "�" in the "raw output"

GoCD 17.2
- non-ASCII characters are replaced by "�" in the "raw output"

Windows Server 2008 R2
Number of pipelines 343
Number of agents 24

 AND 

Windows 7
Number of pipelines 195
Number of agents 12

Firefox Version 56.0
Google Chrome Version 61.0.3163.100
Microsoft Edge 38.14393.1066.0

**Steps to Reproduce**

1. Generate an output containing multiple instances of non-ASCII characters
2. View the console log real time
3. View the raw output

**Expected Results**

Scrolling console log should continue to scroll and correctly display all characters.  Raw output should correctly display all characters.

**Actual Results**

Scrolling console log stops scrolling at the point at which the non-ASCII characters are being processed.  Raw output replaces non-ASCII characters with "�".

**Possible Fix**

Render the characters in the scrolling console log in the same manner that they were rendered in 17.2 and below (I'm not sure exactly where the change took place that caused the problem between versions 17.2 and 17.10).

**Log snippets**

Log is generated normally until this point: (captured from the raw output since the scrolling log truncates prior to displaying this info).  This is just a small portion of the log - there are several occurrences of non-ASCII characters in this log.

![image](https://user-images.githubusercontent.com/17498698/31342370-8ea8e176-acda-11e7-930d-d7a5b5bd9a7f.png)

**Comment by ankitsri11** [comment]
Hi @BrewRunner - I was able to replicate the issue on windows server 16. However, it is not reproducible on MacOS.

One thing which I noticed is that file.encoding value is different in both the case. I suspect this might be causing the problem.

For windows -   "-Dfile.encoding\u003dcp1252"
MacOS - "file.encoding": "UTF-8"

Can you check your http://go_server_url/go/api/support and confirm value for encoding.

**Comment by BrewRunner** [comment]
I do see that same value on both my production system (running Windows Server 2008 and GoCD 17.2) and on my test system (running Windows 7 and GoCD 17.10):

      "-Dfile.encoding\u003dCp1252",

**Comment by BrewRunner** [comment]
Specifying UTF-8 encoding does not change the behavior on my Windows machines.  I started the service with the following in wrapper-server.conf: wrapper.java.additional.16=-Dfile.encoding=UTF-8 and verified in api/support that it was started with encoding set to UTF-8 and the problem persisted.

I then forced all encoding values to UTF-8, again with no change in behavior:
wrapper.java.additional.16=-Dfile.encoding=UTF-8
wrapper.java.additional.17=-Dsun.jnu.encoding=UTF-8
wrapper.java.additional.18=-Dsun.io.unicode.encoding=UTF-8

It appears that the problem lies elsewhere.

**Comment by ankitsri11** [comment]
Upgraded to the latest 17.11 and can see below error logged:

```
2017-10-30 18:49:16,351 ERROR [qtp1405336328-32] ConsoleLogSocket:79 - Session[/0:0:0:0:0:0:0:1:51354:JobIdentifier[testing_3930, 5, 5, defaultStage, 1, defaultJob, 8]:1617202632] closing session because an error was thrown
java.io.UncheckedIOException: java.nio.charset.MalformedInputException: Input length = 1
	at java.io.BufferedReader$1.hasNext(BufferedReader.java:574)
	at java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1811)
	at java.util.Spliterators$1Adapter.hasNext(Spliterators.java:681)
	at com.thoughtworks.go.domain.ConsoleStreamer.stream(ConsoleStreamer.java:59)
	at com.thoughtworks.go.server.websocket.ConsoleLogSender.sendLogs(ConsoleLogSender.java:145)
	at com.thoughtworks.go.server.websocket.ConsoleLogSender.process(ConsoleLogSender.java:88)
	at com.thoughtworks.go.server.websocket.ConsoleLogSocket.onConnect(ConsoleLogSocket.java:66)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.call(CallableMethod.java:70)
	at org.eclipse.jetty.websocket.common.events.JettyAnnotatedEventDriver.onConnect(JettyAnnotatedEventDriver.java:148)
	at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.openSession(AbstractEventDriver.java:227)
	at org.eclipse.jetty.websocket.common.WebSocketSession.open(WebSocketSession.java:421)
	at org.eclipse.jetty.websocket.server.WebSocketServerConnection.onOpen(WebSocketServerConnection.java:72)
	at org.eclipse.jetty.io.AbstractEndPoint.upgrade(AbstractEndPoint.java:185)
	at org.eclipse.jetty.server.HttpConnection.completed(HttpConnection.java:345)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:436)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.nio.charset.MalformedInputException: Input length = 1
	at java.nio.charset.CoderResult.throwException(CoderResult.java:281)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:339)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.io.BufferedReader.readLine(BufferedReader.java:324)
	at java.io.BufferedReader.readLine(BufferedReader.java:389)
	at java.io.BufferedReader$1.hasNext(BufferedReader.java:571)
	... 23 common frames omitted
```

**Comment by ankitsri11** [comment]
I tried on other versions of GoCD (tried with the zip installer on windows box) and could see below behaviour -

17.2 - 17.4 - console log is getting populated without any issue.
17.5 - 17.7 - console log is getting populated but all the special characters are replaced with "�".
17.8 - 17.11 - console log doesn't get populated once the job is finished and displays "waiting for console log..." message. Console is only visible via raw output link or from artifact tab. 

@BrewRunner - If you wish to upgrade, you can try upgrading to 17.7 until we fix this issue.

**Comment by BrewRunner** [comment]
Any chance of this being addressed in an upcoming release?

**Comment by arvindsv** [comment]
Yes, I think we should. 

I believe @maheshp looked at this - and has some findings. Right, @maheshp? Or, am I confusing it with something else?

**Comment by maheshp** [comment]
@arvindsv I was actually looking at https://github.com/gocd/gocd/issues/3930 . Not sure if the changes for it would fix this issue. This issue needs to be looked at though.

@BrewRunner we are close to making a release, it be tough to get it in for this release.

**Comment by arvindsv** [comment]
Ok, let's consider it for the next one then. I'll move it there.

**Comment by ketan** [comment]
@ankitsri11 and @BrewRunner — I think I have a fix for this issue. While this fix has missed the train for the release due tomorrow, this can be merged in time for 18.1.

Can you download the server and agent from #4044 (look for the `installers-PR/dist` pipeline which will contain the installer)

Let me know if it fixes the problem.

**Comment by BrewRunner** [comment]
I am attempting to test this, but a different issue seems to have been introduced that prevents my builds from getting far enough to test the fix.

The GO_P4_CLIENT_[material name] environment variable now contains the server name instead of the agent name.  Historically, the format for the variable has been cruise-[**agent** hostname]-[pipelinename]-[random hash].  Now the format is cruise-[**server** hostname]-[pipelinename]-[random hash].  The historic format containing the agent name is still used when syncing the material from Perforce, but the variable now instead contains the server name, which causes the perforce client to not be found when it is referenced post sync.

**Comment by ketan** [comment]
The name is derived from the `name` attribute on the `<p4/>` tag, and not the agent or server hostname. Can you update the `<p4 name='..'/>` attribute to suit your needs?

**Comment by jyotisingh** [comment]
@BrewRunner - Thanks for testing and pointing it out. [That](https://github.com/gocd/gocd/issues/3926#issuecomment-349350079) looks like a bug with the recent changes that we made around this. We will get that fixed.

@rajiesh, @ketan - need to hold on 17.12 for the issue @BrewRunner just raised.

**Comment by arvindsv** [comment]
@maheshp and I took a quick look at this as well. I guess this needs to be changed in the BuildComposer workflow as well. Was missed there (in https://github.com/gocd/gocd/pull/3405) because this was checked with both agent and server being on the same host, I think.

**Comment by jyotisingh** [comment]
@arvindsv - Logged a separate issue for the issue (https://github.com/gocd/gocd/issues/4058) with build-compose workflow that you pointed out [here](https://github.com/gocd/gocd/issues/3926#issuecomment-349400380)

**Comment by BrewRunner** [comment]
I noticed that this did fix did not make it into 18.1.  Is there a build available containing this fix that I can evaluate?

**Comment by ketan** [comment]
@BrewRunner — sorry for not providing an update earlier. This one is scheduled to be fixed in 18.2, I'll let you know once a build is available that you can test on.

Thank you for volunteering to evaluate.

**Comment by ketan** [comment]
@BrewRunner can you try with a build at https://build.gocd.org/go/tab/build/detail/installers-PR/894/dist/1/dist#tab-artifacts? (Click on the guest login button on the login screen when prompted).

This is an experimental build, so we don't recommend using this on a production server. If you can confirm that this works fine for you, I'll merge that code into the mainline for 18.2.

**Comment by BrewRunner** [comment]
The console log did not stop updating using this experimental build.  The non-ascii characters were replaced with "�", just as they previously have been in the raw output.  Is this is the expected behavior for the change that was made?

**Comment by ketan** [comment]
> The console log did not stop updating using this experimental build. 

Thank you for confirming that.

> The non-ascii characters were replaced with "�", just as they previously have been in the raw output. Is this is the expected behavior for the change that was made?

Can you confirm if the font you see supports unicode characters? I'm testing this on chrome (running on osx) & safari (running on osx)  and IE 11 and IE Edge and things render fine for me.

Screenshots for IE11 and edge:

<img width="998" alt="screen shot 2018-02-07 at 4 21 11 am" src="https://user-images.githubusercontent.com/10598/35888877-57cffc9a-0bbf-11e8-873c-eceb71a189ce.png">

<img width="968" alt="screen shot 2018-02-07 at 4 22 43 am" src="https://user-images.githubusercontent.com/10598/35888882-60a25cdc-0bbf-11e8-9f28-9162601e8291.png">

When you select one of the log lines in chrome dev tools, what value do you see for the font, can you also confirm what browser and OS you're on?

<img width="1436" alt="screen shot 2018-02-07 at 4 26 52 am" src="https://user-images.githubusercontent.com/10598/35888849-46ee6bd2-0bbf-11e8-9e44-764d1ceb7464.png">

**Comment by ketan** [comment]
A fix has been merged, but reopening for some thorough testing

**Comment by BrewRunner** [comment]
I'm using Windows 10 Pro, normally with Firefox Version 58.0.1, but the same behavior is evident using Microsoft Edge 38.14393.1066.0 and Chrome Version 64.0.3282.140.

It looks like Chrome is rendering Consalas for me in both Go Version: 18.2.0 (6025-6c8a9c8407604546b70a8f3e1269ff8a0c2892e4), which **is not** displaying properly, and Go Server Version: 17.2.0 (4587-442a7d67087844e44a2dcf5824164e1dfe00cc45), which **is** displaying properly.

18.2:
![18 2](https://user-images.githubusercontent.com/17498698/35925473-cb3a8978-0bf3-11e8-9378-ddc88edefe35.png)

17.2:
![17 2](https://user-images.githubusercontent.com/17498698/35925762-8b6060d8-0bf4-11e8-9d04-f826edea35dd.png)

**Comment by ketan** [comment]
@naveenbhaskar (our in-house expert on all things UX) has been tasked with identifying a font-face that supports the most unicode range and something we can distribute with gocd, this will avoid having to fallback fonts provided by the OS and allow us the flexibility to update the fonts at a cadence faster than the release cadence of most OSes.

**Comment by ketan** [comment]
I'm going to close this issue as fixed, since we are consistently handling charset encoding across the board. This is currently set to utf-8 and is configurable.

I'm opening https://github.com/gocd/gocd/issues/4487 to track the specific issue about packaging a monospace font that works across the board for multiple languages.
